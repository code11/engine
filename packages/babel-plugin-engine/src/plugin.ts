import * as Babel from "@babel/core";
import { VariableDeclaratorVisitor } from "./visitors";

export type PluginConfig = {
  viewLibrary: string;
};

type pluginType = (babel: typeof Babel, state: PluginConfig) => {};

export const plugin: pluginType = (babel, state) => {
  return {
    name: "babel-plugin-engine",
    visitor: {
      VariableDeclarator: VariableDeclaratorVisitor(babel, state),
    },
  };
};
