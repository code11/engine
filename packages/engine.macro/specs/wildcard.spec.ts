import pluginTester from "babel-plugin-tester";
import prettier from "prettier";
import plugin from "babel-plugin-macros";

const macroFile = "'@c11/engine.macro'";
const config = require("./babelConfig.json");

pluginTester({
  plugin,
  babelOptions: { filename: __filename, ...config },
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: "babel",
    });
  },
  tests: {
    "should support wildcard": {
      code: `
        import { wildcard, view } from ${macroFile};
        const foo: view = ({
          id = wildcard,
          boo = observe.foo[arg.id].foo,
          bar = observe.foo[wildcard].bar
        }) => {}
      `,
      snapshot: true,
    },
  },
});
