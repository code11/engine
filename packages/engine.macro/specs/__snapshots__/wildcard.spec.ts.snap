// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown plugin should support wildcard: should support wildcard 1`] = `

import { wildcard, view } from '@c11/engine.macro';
const foo: view = ({
  id = wildcard,
  boo = observe.foo[arg.id].foo,
  bar = observe.foo[wildcard].bar
}) => {}

      ↓ ↓ ↓ ↓ ↓ ↓

"use strict";

var _engine = require("@c11/engine.producer");

var _engineView = require("@c11/engine-view");

var foo = (0, _engineView.view)({
  props: {
    type: "STRUCT",
    value: {
      id: {
        type: "VALUE",
        value: {
          type: "CONST",
          value: _engine.wildcard,
        },
      },
      boo: {
        type: "OBSERVE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INTERNAL",
            path: ["id"],
          },
          {
            type: "CONST",
            value: "foo",
          },
        ],
      },
      bar: {
        type: "OBSERVE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "CONST",
            value: _engine.wildcard,
          },
          {
            type: "CONST",
            value: "bar",
          },
        ],
      },
    },
    meta: {},
  },
  fn: function fn(_ref) {
    var id = _ref.id,
      boo = _ref.boo,
      bar = _ref.bar;
  },
});


`;
